### PR TITLE
Optional subworkflow task

### PR DESCRIPTION
**Pull request by jxu-nflx**
Pull Request type
----

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes (Please run `./gradlew generateLock saveLock` to refresh dependencies)
- [ ] WHOSUSING.md
- [ ] Other (please describe):

Changes in this PR
----
* Mark JOIN task which depends on an optional task that has COMPLETED_WITH_ERRORS with the same status (i.e. COMPLETED_WITH_ERRORS)
* If subworkflow task is defined as optional in the parent workflow, the parent workflow status will not be updated when the subworkflow is retried/rerun/restarted

_Describe the new behavior from this PR, and why it's needed_
Issue #

Alternatives considered
----

_Describe alternative implementation you have considered_

### CODE REVIEW

File: docs/docs/reference-docs/sub-workflow-task.md [hunk 1]
```diff
@@ -173,3 +173,7 @@ Looking at the subworkflow (the WEBP version):
 
 The ```subWorkflowParam``` tells conductor which workflow to call. The task is marked as completed upon the completion of the spawned workflow. 
 If the sub-workflow is terminated or fails the task is marked as failure and retried if configured. 
+
+### Optional Sub Workflow Task
+If the Sub Workflow task is defined as optional in the parent workflow task definition, the parent workflow task will not be retried if sub-workflow is terminated or failed.
+In additional, even if the sub-workflow is retried/rerun/restarted after reaching to a terminal status, the parent workflow task status will remain as it is.
```
Comment (aravindanr) [inline docs/docs/reference-docs/sub-workflow-task.md:179]: typo `In addition`

File: core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java [hunk 2]
```diff
@@ -653,6 +653,13 @@ private void updateAndPushParents(WorkflowModel workflow, String operation) {
             // update parent's sub workflow task
             TaskModel subWorkflowTask =
                     executionDAOFacade.getTaskModel(workflow.getParentWorkflowTaskId());
+            if (subWorkflowTask.getWorkflowTask() != null
```
Comment (aravindanr) [inline core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java:656]: The null check is redundant, In a TaskModel instance, `WorkflowTask` should always be present.

**Review by aravindanr — APPROVED** [review]

File: core/src/main/java/com/netflix/conductor/core/execution/tasks/Join.java [hunk 3]
```diff
@@ -66,11 +67,26 @@ public boolean execute(
...
+            }
         }
-        if (allDone || hasFailures) {
+        if (allDone || hasFailures || optionalTaskFailures.length() > 0) {
             if (hasFailures) {
                 task.setReasonForIncompletion(failureReason.toString());
                 task.setStatus(TaskModel.Status.FAILED);
+            } else if (optionalTaskFailures.length() > 0) {
+                task.setStatus(TaskModel.Status.COMPLETED_WITH_ERRORS);
+                optionalTaskFailures.append("completed with errors");
```
Comment (apanicker-nflx) [inline core/src/main/java/com/netflix/conductor/core/execution/tasks/Join.java:88]: nit:
```java
optionalTaskFailures.append(" completed with errors");
```

**Review by apanicker-nflx — APPROVED** [review]

File: core/src/main/java/com/netflix/conductor/core/execution/tasks/Join.java (hunk 3 above)
Comment (jxu-nflx) [reply core/src/main/java/com/netflix/conductor/core/execution/tasks/Join.java:88]: There is a space already appended here: https://github.com/Netflix/conductor/pull/2974/files#diff-ebc515b9038973f4691b79e4dbbfec232800c5a87e95cc2ddbd12033b0fd2926R79 so I am omit the space here :)

**Pull request merged**